Add support for port 0 handling in LiveServerTestCase

josephschorr opened the pull request:
Following this change, if the `LIVESERVER_PORT` is set to zero, then the port will be chosen by the operating system and the LiveServerTestCase will adapt accordingly. This requires a bit of a hack (unfortunately), but makes testing large amounts of concurrent LiveServerTestCases much safer.

jcomo commented:
Thanks for the contribution! Good catch on that spelling mistake.

I'm trying to get a better understanding of the port 0 issue. There is a default port if none is specified...are you explicitly setting the port to 0?

josephschorr commented:
In our test suite, we are running multiple LiveServerTestCases, in parallel, via pytest+xdist. This requires that we choose a unique port for each test, across multiple processes. By explicitly specifying a port of `0`, we are guaranteed to get a non-used port by the OS, but LiveServerTestCase needs to adapt to that case, which is what this change does. From a testing perspective, this is actually the "safest" approach.

Unrelated: Looks like `SocketServer` doesn't exist in Python 3; I'll have to do some research on the Python 3 equivalent and update the PR accordingly.

jcomo commented:
Ah I see. Is there no way to set configuration per test run? I understand the argument for letting the OS pick a port as being the most correct way to do it but why not pick a random number in an unused range of ports?

Yes, I noticed this change isn't compatible with python 3, but I'm sure there is an available alternative

josephschorr commented:
We were doing that until recently with an incrementing counter. However, now that we are using pytest+xdist to run tests in parallel, we'd have to create an incrementing counter across processes for every test suite we use. We could do so, but I figured I'd try a solution that works for everyone first :)

jcomo commented:
Just so I understand this correctly...if the liveserver test case had this functionality, you would hard code 0 into your test runs and then call the private `_get_server_address()` to see which port it ends up running on? If that's correct and you're already doing some logic to increment the port and feed it into the test case (which, yes, breaks down in parallel), why not change the logic to just pick a random port value instead?

josephschorr commented:
I send in `0` for the port, but I don't need to call `_get_server_address`. I instead call `get_server_url`, which is the public method on the application, which returns the full path (`localhost:porthere`). Following this change, I don't need to increment or manage the port at all - the setting of `0` does it for me, just like in vanilla Flask:

```py
app.config['LIVESERVER_PORT'] = 0

# Later:
app_url = app.get_server_url() # Contains the proper URL for the running server
```

Random isn't safe - We tried that and were seeing sporadic test failures when either multiple tests picked the same port or a port picked was outside the "allowed" range. Either case is rare, but can happen, and this change negates both possibilities.

jcomo commented:
Alright, I can see the application of this use case. In order to merge these changes in please add/find support for python 3 and this will need to come with appropriate test cases. I think this should also include documentation on how/why to use this feature.

josephschorr commented:
@jcomo: Updated and testing for Python 3. I also changed the unit test for the LiveServerTestCase to make use of the new port 0 functionality, thereby testing it and also making it safer to run concurrently (if you decide to do so in the future)

jcomo commented:
Looking good! I'd prefer to have a separate test for the port 0 case since it is a bit of an edge use case. Also could you add a little note to the docs about that config variable explaining that port 0 works?

josephschorr commented:
@jcomo: Updated tests and docs. Please let me know if the doc updates are sufficient.

jcomo commented on docs/index.rst:
```diff
@@ -77,6 +77,37 @@ PhantomJS you can use the LiveServerTestCase::
+            response = urllib2.urlopen(self.get_server_url())
+            self.assertEqual(response.code, 200)
+
+The method ``get_server_url`` will return http://localhost:{ospickedport}.
```
This line is redundant

josephschorr replied:
Removed

jcomo commented on flask_testing/utils.py:
```diff
@@ -16,6 +16,13 @@
+except ImportError:
+    # Python 2 SocketServer fallback
+    import SocketServer as socketserver
+
```
Only one newline here please :)

josephschorr replied:
Done

jcomo commented:
Looks good! I had a couple minor comments. Once those are cleaned up we'll get it merged. Thanks for the contribution

josephschorr commented:
Updated

jcomo commented:
⚡️
